openweathermap-ios/openweathermap-ios/Weather/Model/WeatherManager.swift

```diff
@@ -0,0 +1,38 @@
+import CoreLocation
+
+protocol WeatherManagerDelegate: class {
+    func didUpdateWeather(weather: WeatherModel)
+}
+
+final class WeatherManager {
+    private let weatherURL = "https://api.openweathermap.org/data/2.5/weather?appid=c6e381d8c7ff98f0fee43775817cf6ad&units=metric"
+    
+    weak var delegate: WeatherManagerDelegate?
+    
+    func fetchWeather(latitude: CLLocationDegrees, longitude: CLLocationDegrees) {
+        let urlString = "\(weatherURL)&lat=\(latitude)&lon=\(longitude)"
+        getRequest(with: urlString)
+    }
+    
+    private func getRequest(with utlString: String) {
+        guard let url = URL(string: utlString) else { return }
+        let session = URLSession(configuration: .default)
+        let task = session.dataTask(with: url) { [weak self] data, _, _ in
+            guard let data = data else { return }
+            if let weather = self?.parseJSON(data) {
+                self?.delegate?.didUpdateWeather(weather: weather)
+            }
+        }
+        task.resume()
+    }
+    
+    private func parseJSON(_ weatherData: Data) -> WeatherModel? {
+        let decoder = JSONDecoder()
+        guard let decodeData = try? decoder.decode(WeatherDataResponse.self, from: weatherData) else { return nil }
+        let cityName = decodeData.name
+        let temp = decodeData.main.temp
+        let description = decodeData.weather[0].description
```

Comment:
Avoid access arrays with an index because it has a good chance to crash.

Reply:
Can this a solution? Or is there another way to solve it?

https://www.wearemobilefirst.com/blog/swift-type-safety-2

Reply:
If you want to access the first element, prefer to use `.first`. It returns the element only if it exists, otherwise it returns nil instead of crashing.

Use `guard let` to protect the creation of the model only if you have all the data you need.

Reply:
`guard let cityName = decodeData.name.first else { return nil }`

For String it worked. But what about Int or Double?

Reply:
That way, only the first character is returning.

We need to take another step. How can I return the string?

![image](https://user-images.githubusercontent.com/21261476/80596248-75d6d980-89fc-11ea-93f3-cc2afe53cf1c.png)

Reply:
Can it be a solution if an array value does not exist?

```
func parseJSON(_ weatherData: Data) -> WeatherModel? {
        let decoder = JSONDecoder()
        do {
            let decodeData = try decoder.decode(WeatherData.self, from: weatherData)
            let cityName = decodeData.name
            let temp = decodeData.main.temp
            let description = decodeData.weather[0].description
            let weather = WeatherModel(city: cityName, temp: temp, description: description)
            return weather
        } catch {
            delegate?.didFailWithError(erro: error)
            return nil
        }
```

Reply:
@henrique-morbin-ifood  Can you help me with this? I do not know what to do.

Reply:
You are accessing the first weather. Replace `[0]` by `.first`.

Don't call the delegate on the `parseJSON` method.

Reply:
I have some problems :D.
This returns a character: `guard let cityName = decodeData.name.first else { return nil }`

Here is the other problem that I don't know how to check if the data exists.
`let temp = decodeData.main.temp`

```
guard let decodeData = try? decoder.decode(WeatherDataResponse.self, from: weatherData) else { return nil }
        guard let cityName = decodeData.name.first else { return nil }
        let temp = decodeData.main.temp
        guard let description: String = decodeData.weather.first?.description else { return nil }
        guard let id = decodeData.weather.first?.id else { return nil }
```

Reply:
Don't use `.first` after Strings. The `cityName` isn't the point here.  

I am talking about the `weather[0]` array. Replace `[0]` by `.first?`. This way you gonna remove the risk of a crash.

In turn, you gonna have an optional `description`. To solve that, do what you already know: `guard let` check.

Create a `WeatherModel` only if the code passes safely by the `decode` and `description` checks. The rest keep as it is.

Reply:
![image](https://user-images.githubusercontent.com/21261476/81130223-502a6280-8f1d-11ea-9e98-fe9679ecdb1a.png)

Even if accessing the matrix like this `[0]` gives an error, I believe that with `.first` it keeps the return of the matrix safe.

It was like this

```
guard let decodeData = try? decoder.decode(WeatherDataResponse.self, from: weatherData) else { return nil }
        let cityName = decodeData.name
        let temp = decodeData.main.temp
        guard let description = decodeData.weather.first?.description else { return nil }
        guard let id = decodeData.weather.first?.id else { return nil }
        return WeatherModel(city: cityName.description, temp: temp, description: description, conditionId: id)
```
Did we finish this part like that?

Reply:
I don't know why you are insisting to keep `[0]` like on the image with the Xcode error. I said more than once to replace it to `.first?`. 

Anyway, the last snippet it's what you need to improve the code and solve this question.